pkg/sentry/socket/plugin/stack/socket.go

```diff
@@ -0,0 +1,874 @@
+	"gvisor.dev/gvisor/pkg/hostarch"
+	"gvisor.dev/gvisor/pkg/marshal"
+	"gvisor.dev/gvisor/pkg/marshal/primitive"
+	"gvisor.dev/gvisor/pkg/safemem"
+	"gvisor.dev/gvisor/pkg/sentry/arch"
+	"gvisor.dev/gvisor/pkg/sentry/fsimpl/sockfs"
+	"gvisor.dev/gvisor/pkg/sentry/inet"
+	"gvisor.dev/gvisor/pkg/sentry/kernel"
+	ktime "gvisor.dev/gvisor/pkg/sentry/kernel/time"
+	"gvisor.dev/gvisor/pkg/sentry/socket"
+	"gvisor.dev/gvisor/pkg/sentry/socket/plugin"
+	"gvisor.dev/gvisor/pkg/sentry/socket/plugin/cgo"
+	"gvisor.dev/gvisor/pkg/sentry/unimpl"
+	"gvisor.dev/gvisor/pkg/sentry/vfs"
+	"gvisor.dev/gvisor/pkg/syserr"
+	"gvisor.dev/gvisor/pkg/usermem"
+	"gvisor.dev/gvisor/pkg/waiter"
+)
+
+type socketOperations struct {
+	vfsfd vfs.FileDescription
+	vfs.FileDescriptionDefaultImpl
+	vfs.DentryMetadataFileDescriptionImpl
+	vfs.LockFD
+	socket.SendReceiveTimeout
+
+	family   int
+	skType   linux.SockType
+	protocol int
+
+	// fd holds the current socket fd created by plugin stack.
+	fd uint32 `state:"nosave"`
+
+	// eventInfo holds current socket fd's corresponding notification queue
+	// and events status. It will be used to interact with plugin
+	// network stack for event reporting.
+	eventInfo plugin.EventInfo
+}
+
+var _ = socket.Socket(&socketOperations{})
+
+const (
+	sizeofSockaddr = syscall.SizeofSockaddrInet6
+
+	// Lo IF index.
+	ifLoIndex = 1
+)
+
+func newSocket(t *kernel.Task, family int, skType linux.SockType, protocol int, notifier *Notifier, fd int, flags uint32) (*vfs.FileDescription, *syserr.Error) {
+	mnt := t.Kernel().SocketMount()
+	d := sockfs.NewDentry(t, mnt)
+	defer d.DecRef(t)
+
+	switch skType {
+	case syscall.SOCK_STREAM:
+		protocol = syscall.IPPROTO_TCP
+	case syscall.SOCK_DGRAM:
+		protocol = syscall.IPPROTO_UDP
+	default:
+		return nil, syserr.ErrSocketNotSupported
+	}
+
+	wq := &waiter.Queue{}
+	sop := &socketOperations{
+		family:    family,
+		fd:        uint32(fd),
+		protocol:  protocol,
+		skType:    skType,
+		eventInfo: plugin.EventInfo{Wq: wq},
+	}
+
+	sop.LockFD.Init(&vfs.FileLocks{})
+
+	vfsfd := &sop.vfsfd
+	if err := vfsfd.Init(sop, linux.O_RDWR|(flags&linux.O_NONBLOCK), mnt, d, &vfs.FileDescriptionOptions{
+		DenyPRead:         true,
+		DenyPWrite:        true,
+		UseDentryMetadata: true,
+	}); err != nil {
+		return nil, syserr.FromError(err)
+	}
+
+	notifier.AddFD(uint32(fd), &sop.eventInfo)
+
+	return vfsfd, nil
+}
+
+// Bind implements socket.Socket.Bind.
+func (s *socketOperations) Bind(t *kernel.Task, sockaddr []byte) *syserr.Error {
+	return int2err(cgo.Bind(s.fd, sockaddr))
+}
+
+// Listen implements socket.Socket.Listen.
+func (s *socketOperations) Listen(t *kernel.Task, backlog int) *syserr.Error {
+	return int2err(cgo.Listen(s.fd, backlog))
+}
+
+// Accept implements socket.Socket.Accept.
+func (s *socketOperations) Accept(t *kernel.Task, peerRequested bool, flags int, blocking bool) (int32, linux.SockAddr, uint32, *syserr.Error) {
+	var (
+		peerAddr       linux.SockAddr
```

Comment:
`peerAddr` isn't delcared or used until the end of the function. Let's just declare it there via:
```go
peerAddr := socket.UnmarshalSockAddr(s.family, peerAddrBuf[:peerAddrlen])
```

Reply:
Thanks for pointing out :) We updated here and other variable declaration to make sure they follow the same convention.